chapter_recurrent-modern/encoder_decoder.ipynb

```diff
@@ -0,0 +1,288 @@
+    "%maven ai.djl.mxnet:mxnet-native-auto:1.7.0-backport"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ai.djl.ndarray.*;\n",
+    "import ai.djl.ndarray.types.Shape;\n",
+    "import ai.djl.ndarray.types.DataType;\n",
+    "import ai.djl.nn.AbstractBlock;\n",
+    "import ai.djl.training.ParameterStore;\n",
+    "import ai.djl.util.PairList;"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public abstract class Encoder extends AbstractBlock {\n",
+    "    /* The base encoder interface for the encoder-decoder architecture. */\n",
+    "    private static final byte VERSION = 1;\n",
+    "\n",
+    "    public Encoder() {\n",
+    "        super(VERSION);\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    abstract protected NDList forwardInternal(\n",
+    "            ParameterStore parameterStore,\n",
+    "            NDList inputs,\n",
+    "            boolean training,\n",
+    "            PairList<String, Object> params);\n",
+    "\n",
+    "    @Override\n",
+    "    public Shape[] getOutputShapes(Shape[] inputShapes) {\n",
+    "        throw new UnsupportedOperationException(\"Not implemented\");\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "hEmJQeipnS4k",
+    "origin_pos": 3
+   },
+   "source": [
+    "## Decoder\n",
+    "\n",
+    "In the following decoder interface,\n",
+    "we add an additional `beginState` function\n",
+    "to convert the encoder output (`encOutputs`)\n",
+    "into the encoded state.\n",
+    "Note that this step\n",
+    "may need extra inputs such as \n",
+    "the valid length of the input,\n",
+    "which was explained\n",
+    "in :numref:`subsec_mt_data_loading`.\n",
+    "To generate a variable-length sequence token by token,\n",
+    "every time the decoder\n",
+    "may map an input (e.g., the generated token at the previous time step)\n",
+    "and the encoded state\n",
+    "into an output token at the current time step."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public abstract class Decoder extends AbstractBlock {\n",
+    "    /* The base decoder interface for the encoder-decoder architecture. */\n",
+    "    private static final byte VERSION = 1;\n",
+    "    public NDArray attentionWeights;\n",
+    "\n",
+    "    public Decoder() {\n",
+    "        super(VERSION);\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    abstract protected NDList forwardInternal(\n",
+    "            ParameterStore parameterStore,\n",
+    "            NDList inputs,\n",
+    "            boolean training,\n",
+    "            PairList<String, Object> params);\n",
+    "\n",
+    "    abstract public NDList beginState(NDList encOutputs);\n",
```

Comment:
should it be initState?

Reply:
Yes, just checked. I must have got confused since RNNModelScratch uses beginState. I'll change it.

Reply:
Done